### PR TITLE
Use dmypy instead of mypy in lintrunner

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118491
* #118481
* #118480
* #118479
* __->__ #118475
* #118469
* #118468
* #118467
* #118432
* #118418
* #118414

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: tools/linter/adapters/mypy_linter.py [hunk 1]
```diff
@@ -122,9 +122,12 @@ def check_files(
     retries: int,
     code: str,
 ) -> List[LintMessage]:
+    # dmypy has a bug where it won't pick up changes if you pass it absolute
+    # file names, see https://github.com/python/mypy/issues/16768
+    filenames = [os.path.relpath(f) for f in filenames]
```
Comment (Skylion007) [inline tools/linter/adapters/mypy_linter.py:127]: absolute filenames would be `os.path.abspath` right? Maybe clarify the comment?

**Comment by Skylion007** [comment]
I am bit confused about the motivation of this change, could you provide some context?

File: tools/linter/adapters/mypy_linter.py (hunk 1 above)
Comment (ezyang) [reply tools/linter/adapters/mypy_linter.py:127]: Err, you must pass RELATIVE paths, not ABSOLUTE paths. The absolute path case is buggy.

**Comment by ezyang** [comment]
Yeah, the goal is to make iterated use of `lintrunner` faster when working on mypy. Obviously, you can do dmypy manually, but many people will not realize this is an option.

**Comment by Skylion007** [comment]
Also dmypy is still experimental to my knowledge, so may want to update mypy before enabling it at the very least to minimize any bugs the pop out.

**Review by suo — APPROVED** [review]
lgtm. dmypy is distributed with mypy?

**Comment by suo** [comment]
From https://mypy.readthedocs.io/en/stable/mypy_daemon.html:

> Each mypy daemon process supports one user and one set of source files, and it can only process one type checking request at a time. You can run multiple mypy daemon processes to type check multiple repositories.

This seems problematic given the lintrunner model of passing only changed files to the linter adapaters.

**Comment by ezyang** [comment]
This is fine, we use dmypy run:

```
The client utility dmypy is used to control the mypy daemon. Use dmypy run -- <flags> <files> to type check a set of files (or directories). This will launch the daemon if it is not running. You can use almost arbitrary mypy flags after --. The daemon will always run on the current host. Example:

dmypy run -- prog.py pkg/*.py
dmypy run will automatically restart the daemon if the configuration or mypy version changes.
```

**Comment by ezyang** [comment]
I will do the mypy upgrade in its own PR

**Pull request closed without merging**